File: LeetCode/51. N-Queens/Solution.java
class Solution {
    public static void main(String[] args) {
        Solution solution = new Solution();

        System.out.println(solution.solveNQueens(5));
    }

    public List<List<String>> solveNQueens1(int n) {
        List<List<String>> result = new ArrayList<>();
        if (n == 0) return result;
        dfs(0, n, new HashSet<>(), new HashSet<>(), new HashSet<>(), result, new ArrayList<>());
        return result;
    }

    private void dfs(int row, int n, Set<Integer> cols, Set<Integer> pie, Set<Integer> na, List<List<String>> results, List<String> tempList) {
        if (row >= n) {
            if (tempList.size() == n) {
                results.add(new ArrayList<>(tempList));
            }
            return;
        }

        for (int col = 0; col < n; col++) {
            if (cols.contains(col) || pie.contains(row + col) || na.contains(row - col)) {
                continue;
            }

            cols.add(col);
            pie.add(row + col);
            na.add(row - col);

            String temp = getStr(n, col);
            tempList.add(temp);
            dfs(row + 1, n, cols, pie, na, results, tempList);

            cols.remove(col);
            pie.remove(row + col);
            na.remove(row - col);
            tempList.remove(temp);
        }
    }

    private String getStr(int n, int col) {
        StringBuilder stringBuilder = new StringBuilder();
        for (int i = 0; i < n; i++) {
            if (i == col) {
                stringBuilder.append("Q");
            } else
                stringBuilder.append(".");
        }
        return stringBuilder.toString();
    }

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> result = new ArrayList<>();
        helper(0, n, new boolean[n], new boolean[2 * n], new boolean[2 * n], new ArrayList<>(), result);
        return result;
    }

    private void helper(int row, int n, boolean[] cols, boolean[] xy_sum, boolean[] xy_diff, List<String> board,
                        List<List<String>> res) {
        if (row >= n) {
            res.add(new ArrayList<>(board));
            return;
        }
        for (int col = 0; col < n; col++) {
            int sum = row + col, diff = row - col + n;
            if (!cols[col] && !xy_sum[sum] && !xy_diff[diff]) {
                char[] rowStr = new char[n];
                Arrays.fill(rowStr, '.');
                rowStr[col] = 'Q';
                String temp = new String(rowStr);
                board.add(temp);
                cols[col] = true;
                xy_sum[sum] = true;
                xy_diff[diff] = true;
                helper(row + 1, n, cols, xy_sum, xy_diff, board, res);
                cols[col] = false;
                xy_sum[sum] = false;
                xy_diff[diff] = false;
                board.remove(temp);
            }
        }
    }
}

File: LeetCode/51. N-Queens/Solution.py
class Solution:
    def solveNQueens1(self, n: int) -> List[List[str]]:
        if n < 1: return []
        self.result = []
        self.cols = set()
        self.pie = set()
        self.na = set()
        self._DFS(n, 0, [])
        return [['.' * col + 'Q' + '.' * (n - col - 1) for col in sol] for sol in self.result]

    def _DFS(self, n, row, cur_state):
        if row >= n:
            self.result.append(cur_state)
            return

        for col in range(n):
            if col in self.cols or (row + col in self.pie) or (row - col in self.na):
                continue

            self.cols.add(col)
            self.pie.add(row + col)
            self.na.add(row - col)

            self._DFS(n, row + 1, cur_state + [col])

            self.cols.remove(col)
            self.pie.remove(row + col)
            self.na.remove(row - col)

    def solveNQueens(self, n: int) -> List[List[str]]:
        def DFS(cols, xy_sum, xy_diff):
            row = len(cols)
            if row == n:
                result.append(cols)
                return
            for col in range(n):
                if (col not in cols) and (row + col not in xy_sum) and (row - col not in xy_diff):
                    DFS(cols + [col], xy_sum + [row + col], xy_diff + [row - col])

        result = []
        DFS([], [], [])
        return [['.' * col + 'Q' + '.' * (n - col - 1) for col in sol] for sol in result]
